python/cuml/dask/common/comms.py

```diff
@@ -0,0 +1,495 @@
+
+        self.worker_addresses = self.get_workers_()
+        self.workers = list(map(lambda x: parse_host_port(x),
+                                self.worker_addresses))
+
+        self.nccl_initialized = False
+        self.ucx_initialized = False
+
+        if comms_p2p and (not is_ucx_enabled() or not has_ucp()):
+            warnings.warn("ucx-py not found. UCP Integration will "
+                          "be disabled.")
+            self.comms_p2p = False
+
+    def get_workers_(self):
+        """
+        Return the list of workers parsed as [(address, port)]
+        """
+        return list(self.client.has_what().keys())
+
+    def worker_ranks(self):
+        """
+        Builds a dictionary of { (worker_address, worker_port) : worker_rank }
+        """
+        return dict(list(zip(self.workers, range(len(self.workers)))))
+
+    def ucp_ports(self):
+        return [(w, self.client.submit(_func_ucp_listener_port,
+                                       self.sessionId,
+                                       random.random(),
+                                       workers=[w]).result())
+                for w in self.workers]
+
+    def worker_ports(self):
+        """
+        Builds a dictionary of { (worker_address, worker_port) : worker_port }
+        """
+        return dict(list(self.ucp_ports()))
+
+    def worker_info(self):
+        """
+        Builds a dictionary of { (worker_address, worker_port) :
+                                (worker_rank, worker_port ) }
+        """
+        ranks = self.worker_ranks()
+        ports = self.worker_ports() if self.comms_p2p else None
+
+        if self.comms_p2p:
+            output = {}
+            for k in self.worker_ranks().keys():
+                output[k] = (ranks[k], ports[k])
+            return output
+
+        else:
+            return ranks
+
+    def create_ucp_listeners(self):
+        """
+        Build a UCP listener on each worker. Since this async
+        function is long-running, the listener is
+        placed in the worker's data dict.
+
+        NOTE: This is not the most ideal design because the worker's
+        data dict could be serialized at any point, which would cause
+        an error. Need to sync w/ the Dask team to see if there's a better
+        way to do this.
+        Ref: https://github.com/rapidsai/cuml/issues/841
+        """
+        [self.client.run(_func_ucp_create_listener, self.sessionId,
+                         random.random(), workers=[w], wait=False)
+         for w in self.worker_addresses]
+
+        self.block_for_init("ucp_listener")
+
+    def stop_ucp_listeners(self):
+        """
+        Stops the UCP listeners attached to this session
+        """
+        self.client.run(_func_ucp_stop_listener,
+                        self.sessionId,
+                        wait=True)
+
+    def block_for_init(self, key):
+
+        # TODO: Add appropriate error handling here to eliminate endless loops
```

Comment:
github issue?

Reply:
This is actually not an issue. The more I dug into it, I realized that any exception thrown would cause the loop to break.